### ISSUE TITLE
Unexpected behavior using `consistent-as-needed` mode with `keyword: true` in `quote-props`

### ISSUE THREAD

**Issue by guerrero** [eslint/eslint#3636]
Rule `quote-props` using `consistent-as-needed` with `keyword` option enabled doesn't work as expected. As [ESLint docs](http://eslint.org/docs/rules/quote-props.html#consistent-as-needed) reflects, using `eslint` with the following code should return an error:

```js
/*eslint quote-props: [2, "consistent-as-needed", {"keywords": true}]*/
var x = {
    while: 1,
    volatile: "foo"
};
```
But instead ESLint validation pass without errors.

I know [how to implement the solution](https://github.com/eslint/eslint/pull/3554#issuecomment-137272544) to this bug and I think I can fix it without problems. But first, following the advice of one of the mantainers of this repo, I've created this issue to let other people review it.

I'm using the latest(`1.3.1`) version of ESLint with only the configuration comment of the example as configuration.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by BYK** [comment]
The issue actually seems larger than that. `consistent-as-needed` doesn't warn when quotes are necessary at all. I'm refactoring the rule to make it use the same mechanism as `as-needed` but it's not as easy as I thought.

In the mean time, feel free to submit a PR if you think you have a solution.
